packages/react/src/JsonFormsContext.tsx

```diff
@@ -126,11 +129,53 @@ const useEffectAfterFirstRender = (
   }, dependencies);
 };
 
+interface JsonFormsCoreStateProps
+  extends Omit<
+    JsonFormsInitStateProps,
+    | 'ajv'
+    | 'data'
+    | 'schema'
+    | 'uischema'
+    | 'validationMode'
+    | 'additionalErrors'
+  > {
+  core: JsonFormsCore;
+  config: Config;
+}
+interface JsonFormsStateProviderProps {
+  children: React.ReactNode;
+  initState: JsonFormsCoreStateProps;
+}
+
 export const JsonFormsStateProvider = ({
   children,
   initState,
   onChange,
-}: any) => {
+}: JsonFormsStateProviderProps & JsonFormsReactProps) => {
+  const onChangeRef = useRef(onChange);
+  useEffect(() => {
+    onChangeRef.current = onChange;
+  }, [onChange]);
```

Comment:
This was copied from the previous code and not introduced by you, but actually we don't need the `useEffect` here. We can simply set the `onChangeRef.current` directly here without the effect.